File: datastruct/tree/trie/src/scala/PrefixTree.scala
import scala.language.postfixOps
import Ordering.Implicits._ // for list comparison

object PrefixTree {
  case class Tree[+K, V] (v: Option[V], cs: List[(List[K], Tree[K, V])]) {
    val value = v
    val children = cs
  }

  def empty[K, V](): Tree[K, V] = Tree(None, List())

  def leaf[K, V](v: V): Tree[K, V] = Tree(Some(v), List())

  def insert[K, V] (t: Tree[K, V], key: List[K], v: V): Tree[K, V] = {
    def ins[K, V] (ts: List[(List[K], Tree[K, V])],
                   key: List[K], v: V): List[(List[K], Tree[K, V])] =
      ts match {
        case List() => List((key, leaf(v)))
        case p :: ps => {
          val (key1, t1) = p
          if (key1 == key) {
            (key, Tree(Some(v), t1.children)) :: ps  // overwrite
          } else if (hasPrefix(key1, key)) {
            branch(key, v, key1, t1) :: ps
          } else {
            p :: ins(ps, key, v)
          }
        }
      }
    Tree(t.value, ins(t.children, key, v))
  }

  def hasPrefix[K] (key1: List[K], key2: List[K]) =
    ! key1.isEmpty && ! key2.isEmpty && key1.head == key2.head

  def branch[K, V] (key1: List[K], v: V,
                    key2: List[K], t2: Tree[K, V]): (List[K], Tree[K, V]) = {
    val key = lcp(key1, key2)
    val m = key.length
    val newKey1 = key1.drop(m)
    val newKey2 = key2.drop(m)
    if (key1 == key) { // insert "an" into "another"
      (key, Tree(Some(v), List((newKey2, t2))))
    } else if (key2 == key) { // insert "another" into "an"
      (key, insert(t2, newKey1, v))
    } else {
      (key, Tree(None, List((newKey1, leaf(v)), (newKey2, t2))))
    }
  }

  // the longest common prefix
  def lcp[K] (xs: List[K], ys: List[K]): List[K] =
    if (!xs.isEmpty && !ys.isEmpty && xs.head == ys.head) {
      xs.head :: lcp(xs.tail, ys.tail)
    } else {
      List()
    }

  // lookup
  def lookup[K, V] (t: Tree[K, V], key: List[K]): Option[V] = {
    def diff[K] (xs: List[K], ys: List[K]) = xs.drop(lcp(xs, ys).length)

    def find[K, V] (assoc: List[(List[K], Tree[K, V])], key: List[K]): Option[V] =
      assoc match {
        case List() => None
        case p :: ps => {
          val (key1, t1) = p
          if (key1 == key) {
            t1.value
          } else if (key.startsWith(key1)) {
            lookup(t1, diff(key, key1))
          } else {
            find(ps, key)
          }
        }
      }
    find(t.children, key)
  }

  def lookupString[V] (t: Tree[Char, V], key: String): Option[V] = lookup(t, key.toList)

  def fromList[K, V] (assoc: List[(List[K], V)]): Tree[K, V] =
    ((empty(): Tree[K, V]) /: assoc) { (t, p) => insert(t, p._1, p._2) }

  def fromStringList[V](assoc: List[(String, V)]): Tree[Char, V] =
    fromList(assoc map (p => (p._1.toList, p._2)))

  def fromString(txt: String): Tree[Char, Int] =
    fromStringList(txt.split("\\s+").toList.zipWithIndex)

  // pre-order travrese to populate keys in lexicographical order
  def keys[K <% Ordered[K], V] (t: Tree[K, V]): List[List[K]] = {
    def keysOfPrefix[K <% Ordered[K], V](prefix: List[K],
                                         t: Tree[K, V]): List[List[K]] = {
      val ts = t.children.sortWith(_._1 < _._1)
      val ks = ts flatMap ((p: (List[K], Tree[K, V])) =>
        keysOfPrefix(prefix ++ p._1, p._2))
      t.value match {
        case None => ks
        case Some(_) => prefix :: ks
      }
    }
    keysOfPrefix(List(), t)
  }

  def stringKeys[K <% Ordered[K], V] (t: Tree[K, V]): List[String] =
    keys(t) map (_.mkString)

  // verification

  // test data
  val assocs = List(List(("a", 1), ("an", 2), ("antoher", 7), ("boy", 3), ("bool", 4), ("zoo", 3)),
                    List(("zoo", 3), ("bool", 4), ("boy", 3), ("another", 7), ("an", 2), ("a", 1)))

  def testBuild(kvs: List[(String, Int)]) = {
    val t = fromStringList(kvs)
    val err = kvs.filter( kv => {
      val (k, v) = kv
      lookupString(t, k) match {
        case Some(x) => x != v
        case None => true
      }
    })
    assert(err.isEmpty, println("err\n" + t.toString))
  }

  def testKeys(kvs: List[(String, Int)]) = {
    val t = fromStringList(kvs)
    val ks1 = stringKeys(t)
    val ks2 = kvs.unzip._1.sortWith(_<_)
    assert(ks1 == ks2, println("ks1=" + ks1.toString + "\nks2=", ks2.toString))
    println("t=" + t.toString + "\nkeys=" + ks1.toString)
  }

  def test() = {
    for (assoc <- assocs) {
      testBuild(assoc)
      testKeys(assoc)
    }
    println("pass " + assocs.size + " tests")
  }
}

File: datastruct/tree/trie/src/PrefixTree.hs
{-
    PrefixTree.hs, Alphabetic Prefix Tree.
    Although it's called 'alphabetic', we abstract the key as sequence
    Copyright (C) 2010, Liu Xinyu (liuxinyu95@gmail.com)

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
-}

module PrefixTree where

import Prelude hiding (lookup)
import Data.List (isPrefixOf, sort, sortBy, inits, nub, find)
import Data.Function (on)
import Control.Arrow (first)
import qualified Data.Map as Map

-- definition
data PrefixTree k v = PrefixTree { value :: Maybe v
                                 , subTrees :: [([k], PrefixTree k v)]}
                      deriving Show

empty = PrefixTree Nothing []

leaf v = PrefixTree (Just v) []

insert [] v (PrefixTree _ ts) = PrefixTree (Just v) ts
insert k v (PrefixTree v' ts) = PrefixTree v' (ins ts) where
    ins [] = [(k, leaf v)]
    ins ((k', t) : ts) | match k k' = (branch k v k' t) : ts
                       | otherwise  = (k', t) : ins ts

match [] _ = True
match _ [] = True
match (a:_) (b:_) = a == b

branch a v b t = case lcp a b of
  (c, [], b') -> (c, PrefixTree (Just v) [(b', t)])
  (c, a', []) -> (c, insert a' v t)
  (c, a', b') -> (c, PrefixTree Nothing [(a', leaf v), (b', t)])

-- longest common prefix
lcp [] bs = ([], [], bs)
lcp as [] = ([], as, [])
lcp (a:as) (b:bs) | a /= b = ([], a:as, b:bs)
                  | otherwise = (a:cs, as', bs') where (cs, as', bs') = lcp as bs

-- lookup
lookup [] (PrefixTree v _) = v
lookup ks (PrefixTree v ts) = case find (\(s, t) -> s `isPrefixOf` ks) ts of
  Nothing -> Nothing
  Just (s, t) -> lookup (drop (length s) ks) t

fromList :: Eq k => [([k], v)] -> PrefixTree k v
fromList = foldr (uncurry insert) empty

fromString :: (Enum v, Num v) => String -> PrefixTree Char v
fromString = fromList . (flip zip [1..]) . words

-- Pre-order traverse to populate keys in lexicographical order.
keys :: Ord k => PrefixTree k v -> [[k]]
keys = keys' [] where
  keys' prefix t = case (value t) of
    Nothing -> kss
    (Just _) -> prefix : kss
    where
      kss = concatMap (\(ks, t') -> keys' (prefix ++ ks) t') ts
      ts = sortBy (compare `on` fst) (subTrees t)

startsWith [] (PrefixTree Nothing ts) = enum ts
startsWith [] (PrefixTree (Just v) ts) = ([], v) : enum ts
startsWith k (PrefixTree _ ts) =
  case find (\(s, t) -> s `isPrefixOf` k || k `isPrefixOf` s) ts of
    Nothing -> []
    Just (s, t) -> [(s ++ a, b) | (a, b) <- startsWith (drop (length s) k) t]

enum = concatMap (\(k, t) -> [(k ++ a, b) | (a, b) <- startsWith [] t])

-- ITU-T keypad (T9) mapping
mapT9 = Map.fromList [('1', ",."), ('2', "abc"), ('3', "def"), ('4', "ghi"),
                      ('5', "jkl"), ('6', "mno"), ('7', "pqrs"), ('8', "tuv"),
                      ('9', "wxyz")]

-- reverse ITU-T keypad map
rmapT9 = Map.fromList $ concatMap (\(d, s) -> [(c, d) | c <- s]) $ Map.toList mapT9

digits = map (\c -> Map.findWithDefault '#' c rmapT9)

-- Given a list of digits, find all candidate words (including partial words)
-- with T9 map from a dictionary (implemented in prefix Tree).
findT9 [] _ = [[]]
findT9 k (PrefixTree _ ts) = concatMap find pfx where
  find (s, t) = map (take (length k) . (s++)) $ findT9 (drop (length s) k) t
  pfx = [(s, t) | (s, t) <- ts, let ds = digits s in
              ds `isPrefixOf` k || k `isPrefixOf` ds]

-- look up the prefix tree up to n candidates
get n k t = take n $ startsWith k t

--
example = insert "zoo" 0 (fromString "a place where animals are for public to see")

-- test data
assocs = [[("a", 1), ("an", 2), ("another", 7), ("boy", 3), ("bool", 4), ("zoo", 3)],
          [("zoo", 3), ("bool", 4), ("boy", 3), ("another", 7), ("an", 2), ("a", 1)]]

-- tests
verify = all (\as ->
                 let t = fromList as in
                   all (\(k, v) -> maybe False (==v) (lookup k t)) as) assocs

verifyKeys = all (\as ->
                   keys (fromList as) == (sort $ fst $ unzip as)) assocs

verifyStartsWith = all verifyLookup [("a", 5), ("a", 6), ("a", 7), ("ab", 2),
                                  ("ab", 5), ("b", 2), ("bo", 5), ("z", 3)]
  where
    lst = [("a", "the first letter of English"),
           ("an", "used instead of 'a' when the following word begins with a vowel sound"),
           ("another", "one more person or thing or an extra amount"),
           ("abandon", "to leave a place, thing or person forever"),
           ("about", "on the subject of; connected with"),
           ("adam", "a character in the Bible who was the first man made by God"),
           ("boy", "a male child or, more generally, a male of any age"),
           ("bodyl", "the whole physical structure that forms a person or animal"),
           ("zoo", "an area in which animals, especially wild animals, are kept so that people can go and look at them, or study them")]
    t = fromList lst
    m = Map.fromList lst
    verifyLookup (k, n) = length r <= n &&
                          all (\(k', v) -> k `isPrefixOf` k' && k' `Map.member` m) r
      where
        r = get n k t

verifyT9 = all verify' $ concatMap (tail . inits) ["4663", "22", "2668437"]
  where
    t9lst = zip ["home", "good", "gone", "hood", "a", "another", "an"] [1..]
    verify' ds = ((==) `on` sort . nub) as bs where
      as = findT9 ds (fromList t9lst)
      bs = filter ((==) ds . digits) (map (take (length ds) . fst) t9lst)

verifyAll = and [verify, verifyKeys, verifyStartsWith, verifyT9]
